### PR TITLE
Draft: Sentence pairs and further stability improvment

### PR DESCRIPTION
**Pull request by franz101**
Added support for:
- Dataloaders with multiple workers by patching the Dataloader class globally
- Added unwatch function for trainer
- Added support for passing the exact classifier layer (input=embeddings, output=logits)
- Added experimental insights context manager

### CODE REVIEW

File: dataquality/integrations/experimental/insights.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    framework = "spacy"
+
+    def enter(self) -> None:
+        spacywatch(self.model)
+
+    def exit(self) -> None:
+        spacyunwatch(self.model)
+
+
+class TrainerInsights(BaseInsights):
```
Comment (Ben-Epstein) [inline dataquality/integrations/experimental/insights.py:70]: maybe add a docstring for future devs that this is for Pytorch Trainers as well as Huggingface

File: dataquality/integrations/experimental/insights.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        self.model = model
+
+    def __enter__(self) -> Any:
+        self.cls.enter()
+        return dq
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self.cls.exit()
+        dq.finish()
+        # return dq.metrics.get_dataframe(self.name, self.run, "train")
```
Comment (Ben-Epstein) [inline dataquality/integrations/experimental/insights.py:127]: 🧹

File: dataquality/integrations/experimental/insights.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            self.name = str(config.current_project_id)
+            self.run = str(config.current_run_id)
+        if len(labels) > 0:
+            dq.set_labels_for_run(labels)
+
+        self.model = model
+
+    def __enter__(self) -> Any:
+        self.cls.enter()
+        return dq
```
Comment (Ben-Epstein) [inline dataquality/integrations/experimental/insights.py:122]: why do we need this?

File: dataquality/integrations/experimental/insights.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+from dataquality import config
+from dataquality.exceptions import GalileoException
+from dataquality.integrations.experimental.keras import unwatch as kerasunwatch
+from dataquality.integrations.experimental.keras import watch as keraswatch
+from dataquality.integrations.spacy import unwatch as spacyunwatch
+from dataquality.integrations.spacy import watch as spacywatch
+from dataquality.integrations.torch import unwatch as ptunwatch
+from dataquality.integrations.torch import watch as ptwatch
+from dataquality.integrations.transformers_trainer import unwatch as trainerunwatch
+from dataquality.integrations.transformers_trainer import watch as trainerwatch
```
Comment (Ben-Epstein) [inline dataquality/integrations/experimental/insights.py:21]: i dont think that we require the user to install all of these libraries (for exmaple Im not sure if keras is required, or spacy). So this may fail. Should we try to be smarter about the imports?

File: dataquality/integrations/torch.py [hunk 5]
```diff
@@ -79,49 +149,68 @@ def _on_step_end(self) -> None:
...
+    :param force_local_patching: Force patching of dataloaders
+    :param unpatch_on_start: Force unpatching of dataloaders
+    instead of global patching
     :return: None
     ```
     dq.log_dataset(train_dataset, split="train")
     train_dataloader = torch.utils.data.DataLoader()
     model = TextClassificationModel(num_labels=len(train_dataset.list_of_labels))
-    watch(model, [train_dataloader,test_dataloader])
+    watch(model, classifier_layer = "classifier")
```
Comment (Ben-Epstein) [inline dataquality/integrations/torch.py:213]: you dont need to set the dataloaders anymore?

File: dataquality/integrations/torch.py [hunk 6]
```diff
@@ -135,24 +224,65 @@ def watch(
     assert dq.config.task_type, GalileoException(
         "dq client must be initialized. " "For example: dq.init('text_classification')"
     )
+    if unpatch_on_start:
+        unwatch(force=True)
+    if not getattr(model, "_dq", False):
```
Comment (Ben-Epstein) [inline dataquality/integrations/torch.py:229]: should this be `if` or `elif`?

File: dataquality/integrations/torch.py [hunk 7]
```diff
@@ -135,24 +224,65 @@ def watch(
...
-            "Parralelization is not yet supported"
-        )
-        dataloader._get_iterator = wrap_fn(  # type: ignore
-            dataloader._get_iterator, patch_iterator_with_store(tl.helper_data)
-        )
+    # Patch the dataloader for passed dataloaders instead of patching the dataloader
+    # class globally
+    if force_local_patching:
+        if len(dataloaders) == 0:
+            raise GalileoException("No dataloaders passed to watch")
```
Comment (Ben-Epstein) [inline dataquality/integrations/torch.py:253]: it seems like you still need the dataloaders (related to [this comment](https://github.com/rungalileo/dataquality/pull/484/files#r1068400063))

File: dataquality/utils/torch.py [hunk 8]
```diff
@@ -91,12 +109,13 @@ def _embedding_hook(
...
             output_detached = output_detached[:, 1:, :]
-        self.helper_data["embs"] = output_detached
+        model_outputs = self.helper_data["model_outputs"]
+        model_outputs["embs"] = output_detached
 
     def _logit_hook(
         self,
         model: Module,
-        model_input: Tensor,
+        model_input: Optional[Tensor],
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:118]: why the change to optional?

File: dataquality/utils/torch.py [hunk 9]
```diff
@@ -106,18 +125,55 @@ def _logit_hook(
...
             logits = logits[:, 1:, :]
-        self.helper_data["logits"] = logits
+        model_outputs = self.helper_data["model_outputs"]
+        model_outputs["logits"] = logits
+
+    def _classifier_hook(
+        self,
+        model: Module,
+        model_input: Any,  # Union[BaseModelOutput, Tensor],
+        model_output: Any,  # Union[Tuple[Tensor], TokenClassifierOutput, Tensor],
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:159]: do the types not work for this one?

File: dataquality/utils/torch.py [hunk 10]
```diff
@@ -223,32 +279,148 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+    """
+
+    def wrap_next_index(func: Callable, key: str = "ids") -> Callable:
+        """
+        Wraps the next index function to store the indices.
+        """
+
+        @wraps(func)
+        def patched_next_index(*args: Any, **kwargs: Any) -> Any:
+            #            print("patched_next_index", args, kwargs)
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:341]: 🧹

File: dataquality/utils/torch.py [hunk 11]
```diff
@@ -223,32 +279,148 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+
+def unpatch(patches: List[Dict[str, Any]] = []) -> None:
+    """
+    Unpatch all patched classes and instances
+    :param patches: list of patches
+    """
+    # unpatch all instances and classes
+    # starting with all classes
+    for patch in patches:
+        # print("unpatching", patch["class"])
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:393]: 🧹

File: dataquality/utils/torch.py [hunk 12]
```diff
@@ -223,32 +279,148 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+        if not hasattr(patch["class"], "_patched"):
+            continue
+        # print("unpatching", patch["class"], patch["attr"])
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:396]: 🧹

File: dataquality/utils/torch.py [hunk 13]
```diff
@@ -223,32 +279,148 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+        delattr(patch["class"], f"_old_{patch['attr']}")
+        delattr(patch["class"], "_patched")
+        # then all instances of the classes found through the garbage collector
+        for obj in gc.get_objects():
+            if (
+                isinstance(obj, patch["class"])
+                and hasattr(obj, "_patched")
+                and getattr(obj, f"_old_{patch['attr']}")
+            ):
+                # print("unpatching gc", obj, patch["attr"])
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:411]: 🧹

File: dataquality/utils/torch.py [hunk 14]
```diff
@@ -223,32 +279,148 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+                setattr(obj, patch["attr"], getattr(obj, f"_old_{patch['attr']}"))
+                delattr(obj, f"_old_{patch['attr']}")
+                delattr(obj, "_patched")
+
+    # If no patched items are passed, unpatch all instances and classes
+    if len(patches) == 0:
+        all_objects = gc.get_objects()
+        for obj in all_objects + [_BaseDataLoaderIter]:
+            if isinstance(obj, _BaseDataLoaderIter) and getattr(obj, "_patched", False):
+                # print("unpatching", obj)
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:421]: 🧹

File: tests/insights/test_insights_hf.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    TrainingArguments,
+)
+
+import dataquality as dq
+from dataquality.integrations.experimental.insights import Insights
+from dataquality.schemas.task_type import TaskType
+from dataquality.utils.thread_pool import ThreadPoolManager
+from tests.conftest import LOCATION
+
+food = load_dataset("sasha/dog-food")
```
Comment (Ben-Epstein) [inline tests/insights/test_insights_hf.py:22]: cant access the internet during tests - can you download this and put it in a tests/assets folder or something similar?

File: tests/insights/test_insights_hf.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+food["test"] = food["test"].map(lambda x, idx: {"id": idx}, with_indices=True)
+
+labels = food["train"].features["label"].names
+label2id, id2label = dict(), dict()
+for i, label in enumerate(labels):
+    label2id[label] = str(i)
+    id2label[str(i)] = label
+
+model_name = "facebook/deit-tiny-patch16-224"
+feature_extractor = AutoFeatureExtractor.from_pretrained(model_name)
```
Comment (Ben-Epstein) [inline tests/insights/test_insights_hf.py:35]: same here, can you add this to an assets folder?

File: tests/insights/test_insights_hf.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        train_dataset=food["train"],
+        eval_dataset=food["test"],
+        tokenizer=feature_extractor,
+    )
+    with Insights(trainer, labels=labels, train_df=train_df, test_df=test_df):
+        trainer.train()
+        trainer.evaluate()
+        ThreadPoolManager.wait_for_threads()
+        assert len(vaex.open(f"{LOCATION}/training/0/*.hdf5")) == len(train_df)
+        assert len(vaex.open(f"{LOCATION}/test/**/*.hdf5")) == len(test_df)
```
Comment (Ben-Epstein) [inline tests/insights/test_insights_hf.py:122]: cant you just add these bottom 2 assetions outside of the `with` block, and then you dont need the `wait_for_threads` function, since the context manager should call `finish` automatically?

File: tests/integrations/hf/hf.py [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+classifier_layer = hm.get_layer_by_name(model, "classifier")
+
+
+def forward_hook(
+    model: Module,
+    model_input: Tensor,
+    model_output: Union[TokenClassifierOutput, Tensor],
+):
+    print("model_output", model_output.shape)
+    print("model_input", model_input[0].shape)
```
Comment (Ben-Epstein) [inline tests/integrations/hf/hf.py:24]: 🧹

File: tests/integrations/hf/hf.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+from torch import Tensor
+from torch.nn import Module
+from transformers import AutoModelForSequenceClassification, AutoTokenizer
+from transformers.modeling_outputs import TokenClassifierOutput
+
+from dataquality.utils.torch import ModelHookManager
+
+# %%
+tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")
+model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased")
```
Comment (Ben-Epstein) [inline tests/integrations/hf/hf.py:13]: ⛔ 🌏

File: tests/integrations/hf/hf.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+# %%
+prediction = model(
+    **tokenizer(["hello world", "beer"], return_tensors="pt", padding=True)
+)
+
+
+# %%
+
+# %%
```
Comment (Ben-Epstein) [inline tests/integrations/hf/hf.py:39]: 🧹

File: tests/integrations/hf/test_cv_hf.py [hunk 21]
```diff
@@ -20,8 +20,8 @@
 from tests.conftest import LOCATION
 
 food = load_dataset("sasha/dog-food")
```
Comment (Ben-Epstein) [inline tests/integrations/hf/test_cv_hf.py:22]: same here

**Review by Ben-Epstein — COMMENTED** [review]
left some comments around.

I dont see a test for the `Insights` call on keras, spacy, or TF

File: dataquality/integrations/experimental/insights.py (hunk 1 above)
Comment (franz101) [reply dataquality/integrations/experimental/insights.py:70]: Absolutely we are going to change the name and functionality will discuss with @quiquelores today. Will move this to a separate branch.

File: dataquality/integrations/experimental/insights.py (hunk 3 above)
Comment (franz101) [reply dataquality/integrations/experimental/insights.py:122]: enter is executed in with.

with x() as dq.
      dq.set_step(1)
we just return the dq client. 

if you want for example log steps

File: dataquality/integrations/experimental/insights.py (hunk 4 above)
Comment (franz101) [reply dataquality/integrations/experimental/insights.py:21]: absolutely.

File: dataquality/integrations/torch.py [hunk 22]
```diff
@@ -48,17 +54,81 @@ def __init__(
...
+            {
+                "ids": [],
+                "last_action": "init",
+                "patches": [],
+                "model_outputs": {},
+                "hook_manager": hm,
+                "model": model,
+            }
+        )
+        return helper_data
```
[the remark this answers is not in the document]
Comment (franz101) [reply dataquality/integrations/torch.py:89]: Agree, looking into it and typing would be cool too

File: dataquality/integrations/torch.py (hunk 5 above)
Comment (franz101) [reply dataquality/integrations/torch.py:213]: YES!

File: dataquality/integrations/torch.py (hunk 6 above)
Comment (franz101) [reply dataquality/integrations/torch.py:229]: 👍

File: dataquality/integrations/experimental/insights.py (hunk 3 above)
Comment (Ben-Epstein) [reply dataquality/integrations/experimental/insights.py:122]: but if you've imported dq at the top 
``import dataquality as dq```

can you use that one? what's the difference?

Comment (franz101) [reply dataquality/integrations/experimental/insights.py:122]: you basically only need a single import but both is supported :)

Comment (franz101) [reply dataquality/integrations/experimental/insights.py]: it's the same just for convenience if you only want a single import

File: tests/insights/test_insights_hf.py (hunk 17 above)
Comment (franz101) [reply tests/insights/test_insights_hf.py:122]: correct. but finish deletes the temp files no?

File: tests/integrations/hf/hf.py (hunk 19 above)
Comment (franz101) [reply tests/integrations/hf/hf.py:13]: good catch

File: tests/integrations/hf/hf.py (hunk 20 above)
Comment (franz101) [reply tests/integrations/hf/hf.py:39]: file was locally to test the forward hook, is removed

File: tests/integrations/hf/test_cv_hf.py (hunk 21 above)
Comment (franz101) [reply tests/integrations/hf/test_cv_hf.py:22]: need some time for the replacement

File: dataquality/utils/torch.py (hunk 8 above)
Comment (franz101) [reply dataquality/utils/torch.py:118]: because the the classifier uses the same hook but does not pass a tensor for the model input. added comment

Comment (Ben-Epstein) [reply dataquality/utils/torch.py:118]: hey please answer my slack

File: tests/integrations/hf/test_cv_hf.py (hunk 21 above)
Comment (franz101) [reply tests/integrations/hf/test_cv_hf.py:22]: I have removed the test because it still uses text classification

File: dataquality/__init__.py [hunk 23]
```diff
@@ -1,6 +1,6 @@
 "dataquality"
 
-__version__ = "v0.8.11"
+__version__ = "v0.8.11a2"
```
Comment (Ben-Epstein) [inline dataquality/__init__.py:3]:
```suggestion
__version__ = "v0.8.12"
```

File: dataquality/integrations/torch.py [hunk 24]
```diff
@@ -48,17 +55,78 @@ def __init__(
...
+        """
+        try:
+            self.hook_manager.attach_classifier_hook(
+                model, self.classifier_hook_with_step_end, classifier_layer
+            )
+        except Exception as e:
+            print(
+                f"Could not attach classifier hook to model. "
+                f"Error: {e}. "
+                f"Please check the classifier layer name: {classifier_layer}"
```
Comment (Ben-Epstein) [inline dataquality/integrations/torch.py:106]: this message seems like things are not going to work. What are you doing below to make it work? 

I would change this to a warning (if things are still OK to proceed), and explain what you are doing instead

File: dataquality/integrations/torch.py [hunk 25]
```diff
@@ -79,49 +147,65 @@ def _on_step_end(self) -> None:
         Log the embeddings, ids and logits.
         :return: None
         """
+        model_outputs = self.helper_data["model_outputs"]
+        # Workaround for multiprocessing
```
Comment (Ben-Epstein) [inline dataquality/integrations/torch.py:151]: can you add more here explaining what is happening and why this works?

File: dataquality/integrations/transformers_trainer.py [hunk 26]
```diff
@@ -113,6 +125,14 @@ def setup(
...
+        elif isinstance(train_dataloader_ds, TorchDataset):
+            item = next(iter(train_dataloader_ds))
+            assert hasattr(item, "keys") and "id" in item.keys(), GalileoException(
+                "Dataset __getitem__ needs to return a dictionary"
+                " including the index id. "
+                'For example: return {"input_ids": ..., "attention_mask":'
+                ' ..., "id": ...}'
+            )
         else:
             raise GalileoException(f"Unknown dataset type {type(train_dataloader_ds)}")
```
Comment (Ben-Epstein) [inline dataquality/integrations/transformers_trainer.py:137]:
```suggestion
            raise GalileoException(f"Unknown dataset type {type(train_dataloader_ds)}. Must be a TorchDataset or Dataset")
```

File: dataquality/integrations/transformers_trainer.py [hunk 27]
```diff
@@ -203,23 +223,40 @@ def on_step_end(
...
+        try:
+            self.hook_manager.attach_classifier_hook(
+                model, self._classifier_hook, classifier_layer
+            )
+        except Exception as e:
+            print(
+                f"Could not attach classifier hook to model. "
+                f"Error: {e}. "
+                f"Please check the classifier layer name: {classifier_layer}"
+            )
```
Comment (Ben-Epstein) [inline dataquality/integrations/transformers_trainer.py:244]: same thing here as above. Maybe this should be a warning, and you should explain what you are doing instead

File: dataquality/integrations/transformers_trainer.py [hunk 28]
```diff
@@ -228,20 +265,43 @@ def watch(
...
+    """
+    [`unwatch`] is used to remove the callback from the trainer
+    :param trainer: Trainer object
+    :return: None
+    """
+    a.log_function("transformers_trainer/unwatch")
+    helper_data = dq.get_model_logger().logger_config.helper_data
+    trainer.remove_callback(helper_data["dqcallback"])
+    trainer._signature_columns = helper_data["signature_cols"]
+    trainer.data_collator = helper_data["orig_collate_fn"]
```
Comment (Ben-Epstein) [inline dataquality/integrations/transformers_trainer.py:307]: ill make a ticket for this, but these keys should be in enums, not strings 

`HelperData.dqcallback` etc

File: dataquality/utils/torch.py [hunk 29]
```diff
@@ -223,32 +281,151 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+        return patched_next_index
+
+    # Store all applied patches
+    if "patches" not in store:
+        store["patches"] = []
+
+    # Patch the dataloader
+    if getattr(_BaseDataLoaderIter, "_patched", False):
+        # logger warning if already patched
+        print("BaseDataLoaderIter already patched")
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:358]: either remove or convert to a warning

File: dataquality/utils/torch.py [hunk 30]
```diff
@@ -223,32 +281,151 @@ def get_embedding_layer_by_name(self, model: Module, name: str) -> Module:
...
+                setattr(obj, patch["attr"], getattr(obj, f"_old_{patch['attr']}"))
+                delattr(obj, f"_old_{patch['attr']}")
+                delattr(obj, "_patched")
+
+    # If no patched items are passed, unpatch all instances and classes
+    if len(patches) == 0:
+        base_dataloaders: List[
+            Union[Type[_BaseDataLoaderIter], _BaseDataLoaderIter]
+        ] = [_BaseDataLoaderIter]
+        for obj in gc.get_objects():
```
Comment (Ben-Epstein) [inline dataquality/utils/torch.py:417]: this might be really slow, no? So many things in the garbage collection

File: tests/integrations/torch/test_text_classification_pt.py [hunk 31]
```diff
@@ -104,10 +108,12 @@ def train(dataloader, model):
     for idx, (label, text, offsets) in enumerate(dataloader):
         optimizer.zero_grad()
         predicted_label = model(text, offsets)
-        loss = criterion(predicted_label, label)
-        loss.backward()
+        # uncommented to speed up testing
+        # loss = criterion(predicted_label, label)
+        # loss.backward()
```
Comment (Ben-Epstein) [inline tests/integrations/torch/test_text_classification_pt.py:113]: are these (and the ones below) meant to be commented out in the official tests?


If we dont want them for testing, should we just remove them?

Comment (franz101) [reply tests/integrations/torch/test_text_classification_pt.py:113]: I think it's nice to have functioning model in tests, but because training is expensive i uncommented it out for testing.

File: dataquality/integrations/transformers_trainer.py (hunk 28 above)
Comment (franz101) [reply dataquality/integrations/transformers_trainer.py:307]: agree 👍

**Review by Ben-Epstein — APPROVED** [review]
🚀
